### ISSUE TITLE
Half of the model is missing

### ISSUE THREAD

**Issue by GeorgeMixas** [Godlander/objmc#14]
Hello, I have this (https://we.tl/t-pjtMWXc144) model that has been decimated through Blender in order to reduce faces, every 3D viewer I have tried displays it with no problems but when converting it to Minecraft more than half the model is missing.

This gets fixed if I triangulate faces in Blender but that adds 3+ times the faces.

To me, it would seem that objmc has a problem converting quads since most of the model is made up of them.

In game: 
![download](https://user-images.githubusercontent.com/57024176/169576951-53556934-e5dd-44ee-aa14-29da4990504f.png)

3D viewer: 
![download](https://user-images.githubusercontent.com/57024176/169577041-d77c53bf-2b41-4980-9c53-2a4d1e246672.png)

**Comment by Godlander** [comment]
i think last time i encountered something similar was due to the normals of the faces being inverted

im not sure how i can fix this

**Comment by Godlander** [comment]
you can try to use recalculate outside normals in blender.

if some faces can't be resolved by that mayb look into this https://blender.stackexchange.com/questions/7910/what-is-non-manifold-geometry

**Comment by GeorgeMixas** [comment]
I thought the same so I made sure that wasn't the case, normally when your faces are inverted you can see the other side of the model but in my case, it's fully transparent, you can enable backface culling in Blender and see that the faces aren't inverted.

Example:
![download](https://user-images.githubusercontent.com/57024176/169591724-b632429b-dde7-40f0-b73b-2df643a0cfce.png)

**Comment by Daminator4113** [comment]
I ran into a similar problem with one of my models, actually it's because the faces aren't modeled correctly. The easiest and fastest way I found to solve the problem is to triangulate all the faces (CTRL+T on blender)

**Model in blender :**
![image](https://user-images.githubusercontent.com/43492665/169601368-3f649164-2bec-451e-8a7b-e9c4e89b5e18.png)

**Model in game :**
![image](https://user-images.githubusercontent.com/43492665/169601543-de383f4c-d698-4d64-a639-a7270963f109.png)

**And it work :**
![image](https://user-images.githubusercontent.com/43492665/169601856-80029567-9624-494c-bde5-a08cfa8e04a0.png)

**Comment by Godlander** [comment]
they didn't want to triangulate the faces bcos it doubles face count.

anyway it's not the faces that are inverted but the normals. did you try recalculating normals?

**Comment by GeorgeMixas** [comment]
I did and then I flipped them just for sanity's sake to make sure that wasn't the case.

@Daminator4113, What did you do for the model to show up like that in Blender? Mine shows up normally but perhaps what you did could be a clue on what's wrong with it, if something is.

Fliped normals: 
![download](https://user-images.githubusercontent.com/57024176/169642764-37f035c9-b203-4181-9710-7fd6a3f2a562.png)

Normal normals:
![download](https://user-images.githubusercontent.com/57024176/169643007-17f9b1ee-2b66-4d1a-971b-8d0bcfddd637.png)

**Comment by Daminator4113** [comment]
If you mean transparency, it's just a visual option :
![image](https://user-images.githubusercontent.com/43492665/169647234-157ca32c-5cbd-4788-a80f-ec62b7937595.png)

And I don't think the problem is from flipped normals, I don't know how to explain, but this glitch appears when the faces is not correct (like my example with a useless vertice).

If you look closely on the weapon model, you will see a lot of incorrect faces, for example :
![image](https://user-images.githubusercontent.com/43492665/169647430-163b236e-42d4-418a-8c64-abe25b04ffc2.png)

**Comment by Godlander** [comment]
> if some faces can't be resolved by that mayb look into this https://blender.stackexchange.com/questions/7910/what-is-non-manifold-geometry

yeah those vertices make normals unresolvable. maybe try the link i sent yesterday to resolve the invalid vertices/edges/faces

**Comment by GeorgeMixas** [comment]
I found the problem, it appears that objmc doesn't like ngons, when I specifically use tris or quads everything works but if some part of the mesh is an ngon Minecraft doesn't render it.

Would it be possible to add support for ngons in the future?

Ngons provide the fewer faces possible with quads second and tris having the most faces as expected, that's also the reason why everything got fixed when I triangulated the faces.
